File: cs/Problems/Day8.cs
namespace AOC2023.Problems;

public class Day8 : IProblem<int>
{
    public int Solve(ReadOnlySpan<char> input) => CountStepsThroughNodes(input);

    private int CountStepsThroughNodes(ReadOnlySpan<char> input)
    {
        Span<Range> lineRanges = stackalloc Range[716];
        var lines = input.Split(lineRanges, InputReader.NewLine);

        var nodeList = new NodeList(size: lines - 2);

        for (int i = 2; i < lines; i++)
            nodeList.AppendNode(input[lineRanges[i]]);

        int steps = 0;
        var instructions = input[lineRanges[0]];

        Node? current = nodeList.GetStartingNode();

        while (current is not null)
        {
            int instruction = steps % instructions.Length;
            current = nodeList.GetNextNode(current!.Value, instructions[instruction]);
            steps++;
        }

        return steps;
    }

    readonly ref struct NodeList(int size)
    {
        static readonly int START_NODE = "AAA".GetHashCode();
        static readonly int END_NODE = "ZZZ".GetHashCode();

        private readonly Dictionary<int, Node> store = new(size);

        public void AppendNode(ReadOnlySpan<char> nodeString) => AppendNode(Node.Parse(nodeString));

        public void AppendNode(Node node) => store.Add(node.Value, node);

        public Node GetStartingNode() => store[START_NODE];

        public Node? GetNextNode(Node current, char direction)
        {
            var nextValue = direction == 'L' ? current.Left : current.Right;
            var next = store[nextValue];
            return next.Value == END_NODE ? null : next;
        }
    }

    readonly record struct Node(int Value, int Left, int Right)
    {
        public static Node Parse(ReadOnlySpan<char> nodeString)
        {
            int value = string.GetHashCode(nodeString[..3]);
            int left = string.GetHashCode(nodeString[7..10]);
            int right = string.GetHashCode(nodeString[12..15]);

            return new(value, left, right);
        }
    }
}

File: ts/problems/day8.ts
import { NEW_LINE } from "../inputReader.js";

const countStepsThroughNodes = (input: string) => {
  const [instructions, _, ...nodeStrings] = input.split(NEW_LINE);
  const nodeList = NodeList.Parse(nodeStrings);

  let steps = 0;
  let node: Node | null = nodeList.GetStartingNode();

  // Wanted to use a recursive function, but set 3 blows up the stack.
  while (node) {
    const instruction = steps % instructions.length;
    node = nodeList.GetNextNode(node, instructions[instruction]);
    steps++;
  }

  return steps;
};

// This one will use classes and stuff...

class NodeList {
  private static START_NODE = "AAA";
  private static END_NODE = "ZZZ";

  #store: Record<string, Node> = {};

  private constructor() {}

  GetStartingNode() {
    return this.#store[NodeList.START_NODE]!;
  }

  GetNextNode(current: Node, direction: string) {
    const nextValue = direction === "L" ? current.left : current.right;
    if (nextValue === NodeList.END_NODE) return null;

    const next = this.#store[nextValue];
    return next;
  }

  static Parse(nodeLists: string[]) {
    const nodeList = new NodeList();
    nodeLists.map(Node.Parse).forEach((node) => nodeList.AddNode(node));
    return nodeList;
  }

  private AddNode(node: Node) {
    this.#store[node.value] = node;
  }
}

class Node {
  value: string;
  left: string;
  right: string;

  constructor(value: string, left: string, right: string) {
    this.value = value;
    this.left = left;
    this.right = right;
  }

  static Parse(nodeString: string) {
    const [value, left, right] = nodeString.match(/[A-Z]{3}/g)!;
    return Object.freeze(new Node(value, left, right));
  }
}

export default countStepsThroughNodes;
